### PR TITLE
Port of the VB Financial module (#31181)

### PR DESCRIPTION
**Pull request by ehasis**
* Add of the Financial module from Reference Source

* Add of some basic tests

### CODE REVIEW

**Comment by cston** [comment]
cc @danmosemsft, @KathleenDollard

**Comment by danmosemsft** [comment]
@ehasis did you take a look at the old test cases in 
https://github.com/dotnet/corefx/compare/master...danmosemsft:old.vb.tests?expand=1
Feel free to borrow from those if not.

**Comment by ehasis** [comment]
> 
> 
> @ehasis did you take a look at the old test cases in
> https://github.com/dotnet/corefx/compare/master...danmosemsft:old.vb.tests?expand=1
> Feel free to borrow from those if not.

That's great, I'll port those tests too.

**Comment by danmosemsft** [comment]
> That's great, I'll port those tests too.

Great. When we've borrowed from those tests, we've done so selectively (some of them aren't good or are duplicative) and we've migrated them to our style (eg using [Theory] as needed, and code formatting/style etc)

**Comment by ehasis** [comment]
> 
> 
> > That's great, I'll port those tests too.
> 
> Great. When we've borrowed from those tests, we've done so selectively (some of them aren't good or are duplicative) and we've migrated them to our style (eg using [Theory] as needed, and code formatting/style etc)

Let me ask you something, the tests are failing because the difference of the rounding of the lasts digits, for example: expected: 57.3268063538875 and actual: 57.32680635388748. What do your recommend, update for the new (more precise) values?

**Comment by danmosemsft** [comment]
>  What do your recommend, update for the new (more precise) values?

We increased the accuracy of some numeric parsing and formatting this release which is almost surely why. So I would update for the new values. 

That means .NET Framework will get different results (these tests should run on both, as that's normally useful to make sure we get the same results on both). In which case you can mark the specific test thus `[SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "Numeric precision differs")]` or similar so it will skip on .NET Framework.

**Comment by danmosemsft** [comment]
If you do that you should eyeball to make sure it's very close to .NET Framework though! Another option is writing code in the test like
```
expected = PlatformDetection.IsFullFramework ? old expected value : new expected value
```

File: src/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs [hunk 1]
```diff
@@ -412,6 +412,24 @@ public sealed class FileSystem
         public static void Write(int FileNumber, params object[] Output) { }
         public static void WriteLine(int FileNumber, params object[] Output) { }
     }
+    [Microsoft.VisualBasic.CompilerServices.StandardModuleAttribute]
+    public sealed class Financial
```
Comment (hughbe) [inline src/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs]: We can make this a `public static` class and removed the internal constructor

Comment (cston) [reply src/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs]: Perhaps, but let's keep this as `sealed class` for consistency with other `Module` definitions in the ref assembly.

**Comment by ehasis** [comment]
In regard of the double/decimal, do you want some action about it? If so, what's the approach do you recommend?

About the tests, as previously talked with @danmosemsft I'll add the calculations based on the cases he linked and also add the non-happy tests.

**Comment by KathleenDollard** [comment]
@ehasis Thank you for this work!

Since the primary purpose of this port is to allow moving of current applications, we're not going to include decimal overloads. Decimal overloads would require more work on boundary conditions since the behavior is different (overflow, divide by zero).

File: src/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs (hunk 1 above)
Comment (ehasis) [reply src/Microsoft.VisualBasic.Core/ref/Microsoft.VisualBasic.Core.cs]: Yes, as @cston said, I followed the the structure of the file to keep consistency.

**Comment by ehasis** [comment]
Not sure what to do from now. The some new tests that I pushed recently are failing only on the "Linux musl_x64_Debug" and I don't have this environment to look for the cause.

**Comment by cston** [comment]
@ehasis, thank you for adding all those test cases! I'll investigate the CI build failure.

**Comment by danmosemsft** [comment]
@dotnet/dnceng the 'Linux x64 Debug' legs on several different PR's are super slow today. It's waiting on Helix. Eg this one's nearly been 2 hours, this one https://dev.azure.com/dnceng/public/_build/results?buildId=256664 has been 2:15 so far. Could you pleae help check whether this is jammed?

**Comment by MattGal** [comment]
@danmosemsft looking presently;  we did have a massive backup yesterday which may be still unclogging

**Comment by danmosemsft** [comment]
BTW I think this meets the current bar (when CI is green) under "•	Blocking adoption for significant number of customers without reasonable workaround" since we know these API have ~3-4% usage in the selected adopters usage category.

**Comment by MattGal** [comment]
@danmosemsft yeah something's on fire for ubuntu.1804.amd64.open;  I'm looking at it now.

**Comment by danmosemsft** [comment]
Thanks for the contribution @ehasis . Perhaps you'd like to look through our other issues marked up for grabs?

**Pull request merged**